### PR TITLE
Add add-unit support for kubernetes models.

### PR DESCRIPTION
**Pull request by Veebers**
## Description of change

Put a veneer over the ```scale-application``` command and allow ```add-unit``` to work with K8s models albeit supporting only the ```num-units``` argument.

## Documentation changes

Added a 'note' in the commands help text although this might need to be expanded.

### CODE REVIEW

File: api/application/client.go [hunk 1]
```diff
@@ -596,8 +596,11 @@ type ScaleApplicationParams struct {
 	// ApplicationName is the application to scale.
 	ApplicationName string
 
-	// Scale is the number of units which should be running.
+	// Scale is the target number of units which should exist
```
Comment (wallyworld) [inline api/application/client.go]: I think "be running." is ok?
If it is changed to exist, end with a "."

File: api/application/client.go [hunk 2]
```diff
@@ -609,6 +612,7 @@ func (c *Client) ScaleApplication(in ScaleApplicationParams) (params.ScaleApplic
 		Applications: []params.ScaleApplicationParams{{
 			ApplicationTag: names.NewApplicationTag(in.ApplicationName).String(),
 			Scale:          in.Scale,
+			ScaleChange:    in.ScaleChange,
```
Comment (wallyworld) [inline api/application/client.go]: We should validate scale args here, just like in.ApplicatioName above

File: apiserver/facades/client/application/application.go [hunk 3]
```diff
@@ -1582,9 +1582,14 @@ func (api *APIBase) ScaleApplications(args params.ScaleApplicationsParams) (para
 		return params.ScaleApplicationResults{}, errors.Trace(err)
 	}
 	scaleApplication := func(arg params.ScaleApplicationParams) (*params.ScaleApplicationInfo, error) {
-		if arg.Scale < 0 {
+		if arg.Scale == 0 && arg.ScaleChange == 0 {
+			return nil, errors.NotValidf("scale of 0")
+		} else if arg.Scale < 0 && arg.ScaleChange == 0 {
 			return nil, errors.NotValidf("scale < 0")
+		} else if arg.Scale != 0 && arg.ScaleChange != 0 {
+			return nil, errors.NotValidf("requesting scale and scale-change")
```
Comment (wallyworld) [inline apiserver/facades/client/application/application.go]: requesting both

File: apiserver/facades/client/application/application.go [hunk 4]
```diff
@@ -1596,6 +1601,14 @@ func (api *APIBase) ScaleApplications(args params.ScaleApplicationsParams) (para
 		} else if err != nil {
 			return nil, errors.Trace(err)
 		}
+		if arg.ScaleChange != 0 {
+			currentScale := app.GetScale()
+			newScale := currentScale + arg.ScaleChange
```
Comment (wallyworld) [inline apiserver/facades/client/application/application.go]: It's not quite this simple as we should add an assert that the current scale hasn't changed. This functionality needs to be moved to state where an assert can be used. ie introduce a ChangeScale() method.

File: cmd/juju/application/addunit.go [hunk 5]
```diff
@@ -35,6 +36,10 @@ with any application or model constraints. This command also supports the
 placement directive ("--to") for targeting specific machines or containers,
 which will bypass application and model constraints.
 
+Note:
+For Kubernetes models the only valid argument is -n, --num-units anything
```
Comment (wallyworld) [inline cmd/juju/application/addunit.go]: full stop after --num-units and Anything on next line

File: cmd/juju/application/addunit.go [hunk 6]
```diff
@@ -184,6 +188,15 @@ func (c *addUnitCommand) Init(args []string) error {
 	if err := cmd.CheckEmpty(args[1:]); err != nil {
 		return err
 	}
+	modelType, err := c.ModelType()
```
Comment (wallyworld) [inline cmd/juju/application/addunit.go]: We won't always know the model type in Init()
(see destroy)

File: cmd/juju/application/addunit.go [hunk 7]
```diff
@@ -184,6 +188,15 @@ func (c *addUnitCommand) Init(args []string) error {
...
+	if err != nil {
+		return err
+	}
+	if modelType == model.CAAS {
+		if c.PlacementSpec != "" || len(c.AttachStorage) != 0 {
+			return errors.New("Kubernetes models only supports num-units flag")
```
Comment (wallyworld) [inline cmd/juju/application/addunit.go]: ... support --num-units"

File: api/application/client.go (hunk 2 above)
Comment (Veebers) [reply api/application/client.go]: I'm not sure what validation makes sense here as remove-unit will pass this a negative number for the ScaleChange. Thoughts?

Comment (wallyworld) [reply api/application/client.go]: No, I meant that we should use the same validation as on the server side, eg both can't be set

File: apiserver/facades/client/application/application_unit_test.go [hunk 8]
```diff
@@ -639,6 +639,58 @@ func (s *ApplicationSuite) TestScaleApplicationsCAASModel(c *gc.C) {
...
+		}},
+	})
+	app := s.backend.applications["postgresql"]
+	app.CheckCall(c, 0, "ChangeScale", 5)
+}
+
+func (s *ApplicationSuite) TestScaleApplicationsCAASModelScaleArgCheck(c *gc.C) {
+	application.SetModelType(s.api, state.ModelTypeCAAS)
+	s.backend.applications["postgresql"].scale = 2
+	results, err := s.api.ScaleApplications(params.ScaleApplicationsParams{
```
Comment (wallyworld) [inline apiserver/facades/client/application/application_unit_test.go]: These would typically be done as a loop
eg see github.com/juju/juju/cmd/juju/subnet/remove_test.go
func (s *RemoveSuite) TestInit(c *gc.C)

File: cmd/juju/application/addunit.go [hunk 9]
```diff
@@ -217,6 +243,28 @@ func (c *addUnitCommand) Run(ctx *cmd.Context) error {
...
+		return err
+	}
+
+	if modelType == model.CAAS {
+		_, err = apiclient.ScaleApplication(application.ScaleApplicationParams{
+			ApplicationName: c.ApplicationName,
+			ScaleChange:     c.NumUnits,
+		})
+		if params.IsCodeUnauthorized(err) {
+			common.PermissionsMessage(ctx.Stderr, "add a unit")
```
Comment (wallyworld) [inline cmd/juju/application/addunit.go]: "scale an application" perhaps

File: state/application.go [hunk 10]
```diff
@@ -1337,6 +1337,39 @@ func (a *Application) GetScale() int {
 	return a.doc.DesiredScale
 }
 
+// ChangeScale alters the existing scale by the provided change amount, returning the new amount.
+// This is used on CAAS models.
+func (a *Application) ChangeScale(scaleChange int) (int, error) {
+	newScale := a.doc.DesiredScale + scaleChange
+	if newScale < 1 {
```
Comment (wallyworld) [inline state/application.go]: newScale < 0

File: state/application.go [hunk 11]
```diff
@@ -1337,6 +1337,39 @@ func (a *Application) GetScale() int {
...
+		return a.doc.DesiredScale, errors.NotValidf("cannot remove more units than currently exist")
+	}
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		if attempt > 0 {
```
Comment (wallyworld) [inline state/application.go]: Need to redo the check that resulting scale is >=0 since if attempt > 0, the scale may have changed underneath us

File: state/application.go [hunk 12]
```diff
@@ -1337,6 +1337,39 @@ func (a *Application) GetScale() int {
...
+				return nil, errors.Trace(err)
+			} else if !alive {
+				return nil, applicationNotAliveErr
+			}
+		}
+		return []txn.Op{{
+			C:  applicationsC,
+			Id: a.doc.DocID,
+			Assert: bson.D{{"life", Alive},
+				{"charmurl", a.doc.CharmURL},
```
Comment (wallyworld) [inline state/application.go]: So this is a bug in Scale() also.
If any of these asserts fail, up above in if attempt > 0, we need to refresh the app to get the latest data, or else we'll just keep looping
			if err := a.Refresh(); err != nil {
				return nil, errors.Trace(err)
			}

File: api/application/client_test.go [hunk 13]
```diff
@@ -1182,6 +1182,41 @@ func (s *applicationSuite) TestScaleApplicationArity(c *gc.C) {
 	c.Assert(err, gc.ErrorMatches, "expected 1 result, got 2")
 }
 
+func (s *applicationSuite) TestScaleApplicationValidation(c *gc.C) {
```
Comment (wallyworld) [inline api/application/client_test.go]: Not a validation test, but we're missing a test for success where scale change is used but not scale. ie a test similar to TestScaleApplication() but for scale change

File: cmd/juju/application/addunit.go [hunk 14]
```diff
@@ -35,6 +36,10 @@ with any application or model constraints. This command also supports the
 placement directive ("--to") for targeting specific machines or containers,
 which will bypass application and model constraints.
 
+Note:
```
Comment (wallyworld) [inline cmd/juju/application/addunit.go]: As with comment in remove-unit PR, I think we need to explain up front that k8s and cloud models are different and preface a block of text for each type.

File: cmd/juju/application/addunit_test.go [hunk 15]
```diff
@@ -66,6 +66,17 @@ func (f *fakeApplicationAddUnitAPI) AddUnits(args apiapplication.AddUnitsParams)
 	return nil, nil
 }
 
+func (f *fakeApplicationAddUnitAPI) ScaleApplication(args apiapplication.ScaleApplicationParams) (params.ScaleApplicationResult, error) {
```
Comment (wallyworld) [inline cmd/juju/application/addunit_test.go]: As with destroy, we need test(s) similar to TestDestroyUnknownModelCallsRefresh etc

File: state/application.go [hunk 16]
```diff
@@ -1337,6 +1337,46 @@ func (a *Application) GetScale() int {
...
+			alive, err := isAlive(a.st, applicationsC, a.doc.DocID)
+			if err != nil {
+				return nil, errors.Trace(err)
+			} else if !alive {
+				return nil, applicationNotAliveErr
+			}
+			if err := a.Refresh(); err != nil {
+				return nil, errors.Trace(err)
+			}
+			newScale = a.doc.DesiredScale + scaleChange
```
Comment (wallyworld) [inline state/application.go]: Something to ponder - I wonder if the scale changes from underneath us, we should just error and not try again. ie if I saw an app had 5 units and I wanted to add 2 more expecting 7, and the scale changed to 4 before my txn ran, the end result would be 6 and not what i expected. Maybe it's better just to error in that case? With a suitable message. Although add-unit has the same problem so maybe not an issue to address right now

File: state/application.go [hunk 17]
```diff
@@ -1337,6 +1337,46 @@ func (a *Application) GetScale() int {
...
+	}
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		if attempt > 0 {
+			alive, err := isAlive(a.st, applicationsC, a.doc.DocID)
+			if err != nil {
+				return nil, errors.Trace(err)
+			} else if !alive {
+				return nil, applicationNotAliveErr
+			}
+			if err := a.Refresh(); err != nil {
```
Comment (wallyworld) [inline state/application.go]: Maybe too subtle - I was hoping for a driveby fix in Scale() also :-)

**Comment by Veebers** [comment]
$$merge$$

File: cmd/juju/application/addunit.go [hunk 18]
```diff
@@ -217,6 +252,28 @@ func (c *addUnitCommand) Run(ctx *cmd.Context) error {
 	}
 	defer apiclient.Close()
 
+	if c.unknownModel {
+		if err := c.validateArgs(); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	modelType, err := c.ModelType()
```
Comment (wallyworld) [inline cmd/juju/application/addunit.go]: Same comment as for remove-unit - I think we pass model type to validate() method.
Also, can we be consistent with naming of the method between add and remove

**Comment by Veebers** [comment]
$$merge$$
